src/twisted/internet/test/test_cfreactor.py

```diff
@@ -0,0 +1,99 @@
+from time import sleep
+from typing import TYPE_CHECKING, List
+
+from twisted.trial.unittest import SynchronousTestCase
+from .reactormixins import ReactorBuilder
+
+if TYPE_CHECKING:
+    fakeBase = SynchronousTestCase
+else:
+    fakeBase = object
+
+
+def noop() -> None:
+    """
+    Do-nothing callable. Stub for testing.
+    """
+
+
+noop()
+
+
+class CoreFoundationSpecificTests(ReactorBuilder, fakeBase):
+    """
+    Tests for platform interactions of the CoreFoundation-based reactor.
+    """
+
+    _reactors = ["twisted.internet.cfreactor.CFReactor"]
+
+    def test_whiteboxStopSimulating(self) -> None:
+        """
+        CFReactor's simulation timer is None after CFReactor crashes.
+        """
+        r = self.buildReactor()
+        r.callLater(0, r.crash)
+        r.callLater(100, noop)
+        self.runReactor(r)
+        self.assertIs(r._currentSimulator, None)
+
+    def test_callLaterLeakage(self) -> None:
+        """
+        callLater should not leak global state into CoreFoundation which will
+        be invoked by a different reactor running the main loop.
+
+        @note: this test may actually be usable for other reactors as well, so
+            we may wish to promote it to ensure this invariant across other
+            foreign-main-loop reactors.
+        """
+        r = self.buildReactor()
+        delayed = r.callLater(0, noop)
+        r2 = self.buildReactor()
+
+        def stopBlocking() -> None:
+            r2.callLater(0, r2stop)
+
+        def r2stop() -> None:
+            r2.stop()
+
+        r2.callLater(0, stopBlocking)
+        self.runReactor(r2)
+        self.assertEqual(r.getDelayedCalls(), [delayed])
+
+    def test_whiteboxIterate(self) -> None:
+        """
+        C{.iterate()} starts the main loop, then crashes the reactor.
+        """
+        r = self.buildReactor()
+        x: List[int] = []
+        r.callLater(0, x.append, 1)
+        delayed = r.callLater(100, noop)
+        r.iterate()
+        self.assertIs(r._currentSimulator, None)
+        self.assertEqual(r.getDelayedCalls(), [delayed])
+        self.assertEqual(x, [1])
+
+    def test_noTimers(self) -> None:
+        """
+        The loop can wake up just fine even if there are no timers in it.
+        """
+        r = self.buildReactor()
+        stopped = []
+
+        def doStop() -> None:
+            r.stop()
+            stopped.append("yes")
+
+        def sleepThenStop() -> None:
+            sleep(0.5)
```

Comment:
What's the point of the sleep?  What order of events is this test trying to exercise?

Reply:
The goal here is to start the loop with no timers, so that scheduleSimulate exercises the "timeout is None" case. If we don't sleep, then the shutdown runs scheduleSimulate with force=True, forcing an immediate shutdown and no further calls to scheduleSimulate. We have to run through startup while fully running the loop, then have some I/O (waker.doRead) wake us up with no timers in sight.

Reply:
Maybe I'm wrong about this. I tried to explain it to myself, and I could not construct the sequence of events that I thought it was there to illustrate.  If codecov agrees then I think we can drop it.

Reply:
OK. I was wrong. The coverage gap I saw on that line must have been for unrelated reasons. The sleep is gone!